File: test/python/bench/spawn_chain.py
"""Mirror of test/jacl/bench/spawn_chain.jacl.

50 sequential spawn+await pairs. asyncio.create_task is the closest
Python analog to JACL's spawn — both submit a unit of work to a
scheduler and return a handle the caller can await. Caveat: asyncio
runs on a single thread and the scheduler is cooperative, so this
measures task-creation + scheduling overhead but not cross-worker
handoff (which JACL's NxM runtime does pay).
"""

import asyncio


async def _unit(i: int) -> int:
    return i + 1


async def _main() -> int:
    acc = 0
    for i in range(50):
        f = asyncio.create_task(_unit(i))
        acc += await f
    return acc


def run() -> int:
    return asyncio.run(_main())


if __name__ == "__main__":
    print(run())

File: test/clojure/bench/spawn_chain.clj
;; Mirror of test/jacl/bench/spawn_chain.jacl.
;;
;; 50 sequential spawn+await pairs. JVM futures via `future` + `@`
;; dispatch onto the agent thread pool; closest semantic to JACL's
;; `spawn` (submit + await). Each future does the same trivial work
;; (i + 1); the cost is dominated by future allocation, task scheduling,
;; completion handoff, and result deref - mirroring spawn_chain's
;; intent of measuring spawn latency rather than throughput.
;; Validates: sum 1..50 = 1275.

(ns bench.spawn-chain)

(defn run []
  (loop [i 0 acc 0]
    (if (< i 50)
      (let [f (future (+ i 1))]
        (recur (inc i) (+ acc @f)))
      acc)))
